fix(elements): allow custom elements to be detached and reattached

lmartorella opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

The commit fixes the issue #38778, allowing a HTML custom element defined via `angular/elements` to be disconnected from the DOM for indefinite time and then re-connected without breaking it.

The current behavior (correctly) destroys the Angular component after 10ms of grace time after receiving the disconnected event. This event can be called if the element is directly disconnected from his parent, or even when a larger DOM tree is disconnected and "parked" in memory for future reattach.

Since the mission of the Angular custom element support is to adhere to the W3C standards and write completely reusable Angular component for interoperability with other frameworks, it can be expected that detached DOM structures that contains a Angular element would be working even after reattaching the tree later on.

Destroying the Angular component when the HTML element is disconnected is correct, since it correctly frees all the resources and allow GC reclaiming of the orphan node.

However the current behavior of the `componentRef.destroy()` is to detach the Element node from his parent. This causes the issue, since the whole tree - when re-connected back later on - won't contain the custom element anymore, causing the issue.

Simply letting the `NgElementImpl` instance to remain attached to his parent mainly fixes the issue of the missing connected callback not called.

Rather than trying to allow a "non-detaching" version of `destroy()`, the proposed fix simply reattach the HTML Custom Element instance back to the DOM in the original position immediately after the `componentRef.destroy()`. This limits the fix to the `elements` package only. I'm not sure if there are some internal utilities to save/restore a node position in the DOM tree, so I used low-level DOM calls to do that.

```ts
          // Save old position of the element node, to maintain it attached even
          // after the this.componentRef.destroy, that detach it from parent
          const attachInfo = this.getElementAttachPoint();

          this.componentRef.destroy();
          this.componentRef = null;

          // Reattach the destroyed empty html element to the parent
          if (attachInfo) {
            const { parent, viewNode, beforeOf } = attachInfo;
            parent.insertBefore(viewNode, beforeOf);
         }          
```

When the node will receive the "connected" event again later, the Angular component should be able to be recreated again from scratch, using the element properties and attributes, exactly like it was the first initialization. The extensive unit test suite already in `elements` is clear about the intent of element reuse, so the fix looks legit.

However, an additional fix is required to let the custom element to not lose the property values. The hosting framework will probably expect the custom element properties to remain in sync, considering the HTML Element implementation as a black box. This is something probably not fully documented in the W3C specs, but when disconnected a custom element should ideally retain his public property values like any other `HTMLElement` instance.

For that, I added a function to copy back input property values from the Component instance to `initialInputValues` immediately before the destruction. I foresee some possible issue about the inputs that uses the `transform` callback.

```ts
  private resetProperties() {
    this.componentFactory.inputs.forEach(input => {
      this.initialInputValues.set(input.propName, this.componentRef!.instance[input.propName]);
    })
  }
```

I don't see any inverted transform function available for inputs, so if we want 100% be sure that the properties are stored back properly the only working solution is to keep the  `initialInputValues` set always in sync with the passed values (the risk in that case is to use more memory for properties that are supposed to only store volatile data decoded on-the-fly by the input).

Last line changed is in `platform-browser`, unfortunately, since the `ShadowDomRendered` must cope with the Element reuse in case of later reconnection. It seems that the W3C specs don't allow `shadowRoot` to be destroyed and then recreated, so the `attachShadow` would raise an exception if called on a element with already a shadow root present. 

The proposed fix there requires a clear of the node content in case of reuse, otherwise duplicated nodes can be created (at least this happened in my tests).

```ts
    if (!this.shadowRoot) {
      this.shadowRoot = (hostEl as Element).attachShadow({mode: 'open'});
    } else {
      // In case of custom elements, it is possible that the host element was already initialized during the
      // element life-cycle (after a disconnect/reconnect)
      this.shadowRoot.innerHTML = "";
    }
```

I've added an additional integration test `reconnect_spec.ts` to test the following cases:

* Direct disconnect/reconnect of the custom element node
* Indirect disconnect/reconnect via disconnection from parent node
* Test of components with shadow DOM and slots.

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No
- [x] Not sure

I don't know how many Angular features can break this fix, since a complete coverage for Angular components should be tested in the `elements` module.

However I don't expect regressions, since the current issue doesn't allow the nodes to be reused at all at the moment.

## Different fix approaches

In order to avoid potential regressions due to the change, the fix could be introduced in different ways:

### Setting

It should be possible to enable/disable such behavior via some setting exposed in the `NgElementConfig`. This would require some little documentation of the current behavior and the new one, to allow backward compatibility.

### Inheritance

The current public `NgElementStrategy` interface is more or less enough to implement an alternative version of `ComponentNgElementStrategy` with the fix, directly in the app.
However this requires ~500 LOC to be copy/pasted in the app, and this is obviously subject to future fixes/evolutions, due to fact that these classes are mainly the core of `angular/elements`.

A viable intermediate solution is to expose the `ComponentNgElementStrategy` in public API as default implementation, to allow:

* override of the destroy behavior
* or exposing flags.

However this solution will require - again - quite extensive documentation of the current internal behavior.

Thanks, L

JeanMeche commented:
Github is reporting some conflicts. Could you please rebase the PR ?

lmartorella commented:
Pushed yesterday, not sure why it is complaining about the commit text (before it was not failing):

```
✘  Invalid commit message
fix(elements): allow custom elements to be detached and reattached
  Error:
    The commit message body does not meet the minimum length of 20 characters
  
  The expected format for a commit is: 
  <type>(<scope>): <summary>
  
  <body>
```

JeanMeche commented:
Currently the comment message _body_ is `Fixes https://github.com/angular/angular/issues/38778` which is 19 char long.

lmartorella committed:
fix(elements): allow custom elements to be detached and reattached

Allows Angular elements to remains detached from DOM and reattached later, without breaking it. Fixes #38778

dgp1130 commented on packages/elements/src/component-factory-strategy.ts:
```diff
@@ -143,14 +144,51 @@ export class ComponentNgElementStrategy implements NgElementStrategy {
+          this.resetProperties();
           this.componentRef.destroy();
           this.componentRef = null;
+          // Reattach the destroyed empty html element to the parent
+          if (attachInfo) {
+            const {parent, viewNode, beforeOf} = attachInfo;
+            parent.insertBefore(viewNode, beforeOf);
+          }
```
Question: Don't we need the element to remain disconnected to adhere to the expected contract of `disconnectedCallback`?

Maybe I'm misunderstanding, but I feel like this would result in:

```typescript
const someAngularElement = document.querySelector('my-element');
someAngularElement.remove();

console.log(someAngularElement.isConnected); // false

await timeout(10);

console.log(someAngularElement.isConnected); // true
```

Even if the element isn't left on the DOM, attaching and detaching it would still be observable from triggering another round of `connectedCallback` / `disconnectedCallback`, so I'd be hesitant to land a fix here which alters the lifecycle in an observable way.

lmartorella replied:
Mainly the issue with the Angular Elements lifecycle is that, when disconnected from the DOM, his inner `componentRef.destroy()` is called. For some reason, the Angular destroy function is detaching the DOM host element from his parent as part of the destroy process, creating let's say a "zombie" HTML Custom Element.

Then if such custom element is reattached to the DOM via a whole fragment reattach, the element won't receive the attach event anymore and remains stale.

More description in the body of issue.

> Don't we need the element to remain disconnected to adhere to the expected contract of disconnectedCallback?

The HTML element must remain attached to his parent, in case an ancestor is detached from the DOM (all nodes receive the detached event) and the whole tree is still referred and ready to be reattached as a whole.

> Even if the element isn't left on the DOM, attaching and detaching it would still be observable from triggering another round of connectedCallback / disconnectedCallback,

Such events are only raised when the element will be re-attached, even indirectly (via ancestor attaching), to the DOM. If a custom element is directly removed from his parent it won't be part of the tree anymore, so not indirectly attached again.

dgp1130 commented on packages/elements/src/component-factory-strategy.ts:
```diff
@@ -143,14 +144,51 @@ export class ComponentNgElementStrategy implements NgElementStrategy {
+   */
+  private resetProperties() {
+    this.componentFactory.inputs.forEach((input) => {
+      this.initialInputValues.set(input.propName, this.componentRef!.instance[input.propName]);
```
Bug: I think we need to call `getInputValue` and potentially unbox from any signals: https://github.com/angular/angular/blob/d3e76b195608aa91d35f7b18923b110db88152d1/packages/elements/src/create-custom-element.ts#L243-L244

I don't think property names are guaranteed to match input names due to Closure compiler property renaming, so we probably need to use only input names here.

dgp1130 commented on packages/elements/test/reconnect_spec.ts:
```diff
@@ -0,0 +1,196 @@
+  beforeAll((done) => {
+    testContainer = document.createElement('div');
+    document.body.appendChild(testContainer);
+    destroyPlatform();
```
Nit: There's already a `destroyPlatform` in the `afterAll`, I don't think we should need one here?

lmartorella replied:
Sorry, I copy-pasted from another test, probably `create-custom-element_spec.ts`, I'm not an expert of Angular test rigs.

dgp1130 commented on packages/elements/test/reconnect_spec.ts:
```diff
@@ -0,0 +1,196 @@
+          customElements.define(compFactory.selector, createCustomElement(comp, {injector}));
+        });
+      })
+      .then(done, done.fail);
```
Style: Prefer `async` / `await` for managing asynchronous delays.

dgp1130 commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -522,7 +522,14 @@ class ShadowDomRenderer extends DefaultDomRenderer2 {
+    } else {
+      // In case of custom elements, it is possible that the host element was already initialized during the
+      // element life-cycle (after a disconnect/reconnect)
+      this.shadowRoot.innerHTML = '';
```
Question: What content is this clearing? Shouldn't the rendered component template be removed as part of `ComponentRef.prototype.destroy`? It feels weird to empty out the shadow root at this location, as I feel like it should already be empty and we just need to not re-create the shadow root.

lmartorella replied:
Yeah, it feel strange that the content is still there.
Probably a better way to implement the fix is to check why the destroy is leaving the shadow DOM intact, but this probably has a bigger impact than the proposed fix, that only runs in case of element re-attaching.

lmartorella committed:
- UT linted

lmartorella committed:
- Fixed getting input values back

dgp1130 reviewed (commented):
Looking at this a little closer, I see two potential problems here:

First, I'm not convinced it's a good idea to destroy and recreate the component, given that it drops all internal component state. We can preserve custom element properties in the wrapper custom element, but any local `signal` state within a component is just gone. You may as well throw away the entire element and just make a new one from scratch. That's not to say that the ultimate goal of this PR isn't valuable, you shouldn't _have_ to recreate the element based on the custom elements contract. I'm just pointing out that the result of this PR isn't actually as useful on its own as it might appear.

Losing internal component state is arguably a flaw in the existing architecture, and in that sense, this PR is a bug fix which just reveals an existing incorrect behavior, but in practice it really just moves the bug from one place to another and I'm not _entirely_ convinced it's an improvement. I worry that we're shifting from one set of incorrect semantics (reattach after 10ms doesn't activate the component) to another set of incorrect semantics (reattach after 10ms resets internal component state).

I filed an issue with a more concrete description of the problem and a wider proposal for how to fix it: https://github.com/angular/angular/issues/64229

Second, I do think [George's argument about content projection](https://github.com/angular/angular/issues/38778#issuecomment-690433315) is valid, and I think this is fundamentally an architectural flaw in the way content projection is used here. Once we've rendered the element once, we've effectively _moved_ any child nodes into the projected `<ng-content>` location, meaning that a future reattach will no longer find those same nodes in the host element's children and can't replicate content projection in the same manner. I suspect this PR fails when attempting to reattach Angular Elements containing projected content.

He suggested caching the nodes, which would probably work with the existing model, but again surfaces a conceptual problem with that model in that you can't change a custom element's children after the initial render. It should be feasible to render with a `<div>` projected and then later remove that `<div>` and change it to a `<span>`. There's semantics for this in custom elements (remove the old child and add a new one), but this not possible for Angular Elements because we're 1) moving the originally projected children and 2) don't react to new children being added.

Again, we can kind of interpret this PR as a partial bug fix revealing other bugs, but I'm hesitant of the https://www.hyrumslaw.com/ implications here where we're moving from a state which is consistently broken and not useful to one which is inconsistently broken and partially useful such that developers might not realize when they're relying on the broken behavior and actually make it even _harder_ to fix properly in the long term.

I also filed an issue with more details on this front: https://github.com/angular/angular/issues/64228

If my above understanding is correct, then my current intuition is that this PR doesn't quite pass the cost/benefit analysis. I suspect the churn and risk of instability likely outweighs the incremental improvement this brings given the limited impact as a result of existing architectural decisions in Angular Elements.
